### ISSUE TITLE
[Maybe an issue?] No inter-sample variation of per-step attention masks

### ISSUE THREAD

**Issue by zmjjmz** [titu1994/tf-TabNet#9]
I was trying to get a sense of feature importance from TabNet using the attention masks, and essentially what I _want_ to do is show the distribution over the sample axis of the attention paid to each feature per step. For a first approximation of that I'm plotting the mean attention paid w/the standard deviation as an error bar, and I noticed that the standard deviation is nearly zero!

![Screenshot from 2020-06-24 14-13-49](https://user-images.githubusercontent.com/1694612/85610475-0ad50800-b625-11ea-9173-958d9e8e7eea.png)

This is the attention masks after running inference on the first batch in the included Iris example, and a text-only version can be found in [this gist](https://gist.github.com/zmjjmz/dcfcb889c88115347de691b69067bff1). The little lines on top of each bar are supposed to be error-bars. I've observed this on a larger dataset with proprietary features (that I can't share here), so it's not just an issue with there being only 50 examples in that inference batch.

So first off - am I [interpreting](https://gist.github.com/zmjjmz/dcfcb889c88115347de691b69067bff1#file-train_iris-py-L13) these masks correctly? Secondly, am I wrong in assuming there even _should_ be inter-sample variation here? Or is the model supposed to apply the same attention to each feature in each step, regardless of the sample? I'm not clear on that -- since it seems like the actual feature value should affect the `transform_coef` block that feeds the `mask_values`, but maybe I'm misunderstanding this.

FWIW, tested a few things (reducing sparsity coefficient, using batch norm) and I haven't found anything that changes this behavior. I haven't tried the original implementation to see if it exhibits similar behavior, but if it's not supposed to behave like this that would be the next step.

**Comment by zmjjmz** [comment]
Ok, I finally got the google-research model working and can confirm that it seems like this behavior isn't expected. 

While this forest covertype model (I couldn't be bothered to use the iris data or my own here, so hopefully that's not causing this issue) isn't fully trained due to me not wanting to wait a day for that to happen, it is trained _somewhat_ and gets about 60% validation accuracy. As I write this I'm training it for more steps to see if this changes when the model is more converged. 

That same plot as above looks like (for the features that get the most attention):
![Screenshot from 2020-06-30 22-02-40](https://user-images.githubusercontent.com/1694612/86194799-73d4e800-bb1d-11ea-93a0-a8bc30c4d2d1.png)

Notably there isn't much _visible_ variation per step, but I looked more closely at it and it appears there is variation, just not super visible.

So, for example, with the google-research implementation the attention distribution over the samples looks like (for feature 43, step 1) 
![Screenshot from 2020-06-30 22-07-58](https://user-images.githubusercontent.com/1694612/86195115-32910800-bb1e-11ea-9d23-b8bd4a018d6c.png)

While it's mostly not getting any attention, there's clearly samples where it gets more attention. Compared to a similar graph for this implementation's attention mask on the Iris model, we get uh

![Screenshot from 2020-06-30 22-12-04](https://user-images.githubusercontent.com/1694612/86195374-c367e380-bb1e-11ea-9057-e3e6fdee761e.png)

You can't tell because I'm not hovering over it, but that's all `0.29` thereabouts.

I see a similar pattern w/this implementation for my own data, which has a much larger amount of samples! I originally thought that maybe the sparsity lead to the feature being zero most of the time, but then realized that the features aren't zero most of the time, they're some non-zero value all of the time.

So, I think this is clearly an issue. I'm training the google-research model further to better investigate this, and I'll hunt around the code a bit to see if there's something obvious going on!

**Comment by titu1994** [comment]
Sorry for the late response, seemingly didn't get notified about the earlier post. 

This is an important problem. I'll take a look at this as well, but we've both gone over the code a few times on both implementations. Can't seem to figure it out. I'll try again this weekend to see what's going on. Thanks for bringing up this point !

**Comment by zmjjmz** [comment]
Hey @titu1994 -- is there any update on this?

**Comment by titu1994** [comment]
Wasn't able to fix it yet. I don't get the difference, the code looks to be very close to the main version

**Comment by zmjjmz** [comment]
Yeah I agree, it's bizarre. I'll see if I can spend some time banging my head against it this week.

**Comment by astrogilda** [comment]
any updates on this guys?

**Comment by zmjjmz** [comment]
I haven't had a chance to look at this due to a shift in priorities on my end

**Comment by titu1994** [comment]
@zmjjmz I've pushed a fix for the iris dataset. Older TF were silently allowing a [Batchsize, 0] (what the heck even is this in reality) tensor as output of https://github.com/titu1994/tf-TabNet/blob/master/tabnet/tabnet.py#L267 step. This raised issues in newer TF and cuda versions, but since I update quite slowly, this silently remained as a bug. Even after this patch though, seems there was not much improvement in std.

Then I changed your script above to use `virtual_batch_size` which seems to have significantly changed the values of std in your script. I haven't managed to plot it due to lack of time, but could you give it a shot?

Simply change the model in your script to the below.
```python
model = tabnet.TabNetClassifier(feature_columns, num_classes=3,
                                feature_dim=8, output_dim=4,
                                num_decision_steps=4, relaxation_factor=1.0,
                                sparsity_coefficient=1e-5, batch_momentum=0.98,
                                virtual_batch_size=25, norm_type='batch',
                                num_groups=1)
```

**Comment by zmjjmz** [comment]
Can confirm that for the iris example I'm now seeing it fixed with the latest `tabnet` package! 
```
   step       feature mean_attention std_attention
2     0  petal_length     0.16009106     0.3665669
6     1  petal_length    0.099461034     0.1591588
10    2  petal_length     0.26774257    0.18403427
3     0   petal_width     0.19118004     0.3713626
7     1   petal_width      0.1508041    0.22828263
11    2   petal_width     0.24146645    0.37346786
0     0  sepal_length     0.10968613    0.23172563
4     1  sepal_length     0.21755588    0.15963018
8     2  sepal_length     0.19437219    0.16577306
1     0   sepal_width      0.5390427    0.47026247
5     1   sepal_width       0.532179    0.22574042
9     2   sepal_width     0.29641876    0.25073248
```
Not sure if this issue was specific to the iris example, but I'll try it out on my dataset and update if there's still an issue.

**Comment by titu1994** [comment]
Ah it's necessary cause the features_for_coeff step slices the feature dimension in the syntax [:, self.output_dim:]. However this feature dim was originally the size of [:, self.output_dim]. So the the final slice on older versions of tensorflow returns a tensor [:, 0] which technically doesn't make sense. Newer tensorflow raises a weird work 0 > 0 error on GPUs cause of Cuda, but still somehow runs on CPUs (don't even ask how that works, TF is weird sometimes). So now due to the assertion, the feature_dim > outout_dim, so feature_dim - output_dim > 0 always, so at least 1 feature will be used for the remaining steps in the loop

**Comment by zmjjmz** [comment]
Of course it can't be simple! I've confirmed that I'm now seeing inter-sample variation in attention masks on my dataset, so I'll go ahead and close this! Thanks!

**Comment by titu1994** [comment]
Thanks ! And I apologise for the huge delay, this should have been an easy fix if I just updated more regularly.
